File: src/tools/shared/resourceLimits.ts
import type { DraftServiceSpec } from "./specSchemas";

export const MAX_SERVICES_PER_STACK = 25;
export const VALID_PORT_RANGE = { min: 1, max: 65535 } as const;
export const PRIVILEGED_PORT_THRESHOLD = 1024;

export interface ResourceLimitIssue {
  code: "too_many_services" | "invalid_port" | "privileged_port";
  path: string;
  message: string;
}

export function checkResourceLimits(
  services: Record<string, DraftServiceSpec>,
): ResourceLimitIssue[] {
  const issues: ResourceLimitIssue[] = [];

  const names = Object.keys(services);
  if (names.length > MAX_SERVICES_PER_STACK) {
    issues.push({
      code: "too_many_services",
      path: "services",
      message: `stack has ${names.length} services; maximum is ${MAX_SERVICES_PER_STACK}`,
    });
  }

  for (const [svcName, spec] of Object.entries(services)) {
    for (let i = 0; i < (spec.ports ?? []).length; i++) {
      const portValue = spec.ports?.[i];
      if (portValue === undefined) continue;
      const hostPort = extractHostPort(portValue);
      if (hostPort === null) continue;
      const path = `services.${svcName}.ports[${i}]`;
      if (hostPort < VALID_PORT_RANGE.min || hostPort > VALID_PORT_RANGE.max) {
        issues.push({
          code: "invalid_port",
          path,
          message: `host port ${hostPort} is outside valid range ${VALID_PORT_RANGE.min}-${VALID_PORT_RANGE.max}`,
        });
        continue;
      }
      if (hostPort < PRIVILEGED_PORT_THRESHOLD) {
        issues.push({
          code: "privileged_port",
          path,
          message: `host port ${hostPort} is privileged (< ${PRIVILEGED_PORT_THRESHOLD}); use a port >= ${PRIVILEGED_PORT_THRESHOLD}`,
        });
      }
    }
  }

  return issues;
}

function extractHostPort(value: string): number | null {
  const trimmed = value.trim();
  const slash = trimmed.lastIndexOf("/");
  const body = slash >= 0 ? trimmed.slice(0, slash) : trimmed;
  const parts = body.split(":");
  if (parts.length === 1) return null;
  const hostSegment = parts.length === 2 ? parts[0] : parts[parts.length - 2];
  if (hostSegment === undefined) return null;
  const match = hostSegment.match(/^(\d+)$/);
  if (!match) return null;
  return Number(match[1]);
}

File: src/tools/shared/resource_limits.py
"""Stack resource and port limit checks.

Parity: ``src/tools/shared/resourceLimits.ts``.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from src.tools.shared.spec_schemas import DraftServiceSpec

MAX_SERVICES_PER_STACK = 25
VALID_PORT_RANGE = {"min": 1, "max": 65535}
PRIVILEGED_PORT_THRESHOLD = 1024


@dataclass
class ResourceLimitIssue:
    code: str
    path: str
    message: str


def _extract_host_port(value: str) -> int | None:
    trimmed = value.strip()
    slash = trimmed.rfind("/")
    body = trimmed[:slash] if slash >= 0 else trimmed
    parts = body.split(":")
    if len(parts) == 1:
        return None
    host_segment = parts[0] if len(parts) == 2 else parts[-2]
    match = re.match(r"^(\d+)$", host_segment)
    if not match:
        return None
    return int(match.group(1))


def check_resource_limits(
    services: dict[str, DraftServiceSpec],
) -> list[ResourceLimitIssue]:
    """Check service count and host port ranges."""
    issues: list[ResourceLimitIssue] = []
    names = list(services.keys())
    if len(names) > MAX_SERVICES_PER_STACK:
        issues.append(
            ResourceLimitIssue(
                code="too_many_services",
                path="services",
                message=(
                    f"stack has {len(names)} services; "
                    f"maximum is {MAX_SERVICES_PER_STACK}"
                ),
            )
        )

    for svc_name, spec in services.items():
        for index, port_value in enumerate(spec.ports or []):
            host_port = _extract_host_port(port_value)
            if host_port is None:
                continue
            path = f"services.{svc_name}.ports[{index}]"
            if (
                host_port < VALID_PORT_RANGE["min"]
                or host_port > VALID_PORT_RANGE["max"]
            ):
                issues.append(
                    ResourceLimitIssue(
                        code="invalid_port",
                        path=path,
                        message=(
                            f"host port {host_port} is outside valid range "
                            f"{VALID_PORT_RANGE['min']}-{VALID_PORT_RANGE['max']}"
                        ),
                    )
                )
                continue
            if host_port < PRIVILEGED_PORT_THRESHOLD:
                issues.append(
                    ResourceLimitIssue(
                        code="privileged_port",
                        path=path,
                        message=(
                            f"host port {host_port} is privileged "
                            f"(< {PRIVILEGED_PORT_THRESHOLD}); "
                            f"use a port >= {PRIVILEGED_PORT_THRESHOLD}"
                        ),
                    )
                )
    return issues


__all__ = [
    "MAX_SERVICES_PER_STACK",
    "PRIVILEGED_PORT_THRESHOLD",
    "VALID_PORT_RANGE",
    "ResourceLimitIssue",
    "check_resource_limits",
]